File: Memory/RAM.py

from Memory import Memory

class RAM(Memory):

   def __init__(self, latency = 100, burst = 0):
      self.latency = latency
      self.burst = burst

   def __str__(self):
      result  = '(ram '
      result += '(latency ' + str(self.latency) + ')'
      if self.burst != 0:
         result += '(burst ' + str(self.burst) + ')'
      result += ')'
      return result

   def process(self, write, addr, size):
      assert(size > 0)
      word_size = self.machine.word_size
      offset = addr % word_size
      count = (size + word_size + offset - 1) // word_size
      if self.burst == 0:
         t = count * self.latency
      else:
         t = self.latency + self.burst * (count - 1)
      self.machine.time += t
      return 0

File: fastsim/ram.ml
open Main_memory

class ram =
    object (self)
        inherit main_memory as super

        val mutable word_size = 4
        val mutable latency = 100

        method set name value =
            match name with
            | "word_size" -> word_size <- int_of_string value
            | "latency" -> latency <- int_of_string value
            | _ -> super#set name value

        method energy t = 0.0

        method word_size = word_size

        method private process base start write addr size =
            writes <- writes + (if write then 1 else 0);
            reads <- reads + (if write then 0 else 1);
            start + latency

    end
